### PR TITLE
Add support for async functions that have a non-trivial ..

### LINKED ISSUE #535: Add support for async functions that have a non-trivial return type of the finish() function
Lots of GIO functions actually return the result instead of having it as an out parameter. Fixing this would remove the need of lots of manual implementations.

### PR DESCRIPTION
**Pull request by MathieuDuponchelle**
return type of the finish() function

Fixes #535 

I've got a WIP patch for Gio here: https://github.com/MathieuDuponchelle/gio/commit/babfd3a287f4050227b681aa45946c4fc95482ac

I also fixed a bug in the same commit, where we would look up the wrong finish function in gio, because we looked it up by rust name, and there were multiple unrelated functions with that name (connect_finish, maybe others). It was a bit awkward to split up, just let me know if you'd rather I split it up!

### CODE REVIEW

File: src/analysis/bounds.rs [hunk 1]
```diff
@@ -117,9 +118,13 @@ impl Bounds {
             if let Some(bound_type) = Bounds::type_for(env, par.typ, par.nullable) {
                 ret = Some(Bounds::get_to_glib_extra(&bound_type));
                 if async && par.name == "callback" {
-                    let finish_func_name = replace_async_by_finish(&func.name);
+                    let func_name = func.c_identifier.as_ref().unwrap();
+                    let finish_func_name = replace_async_by_finish(&func_name);
```
Comment (EPashkin) [inline src/analysis/bounds.rs]: Unneeded &

File: src/codegen/function_body_chunk.rs [hunk 2]
```diff
@@ -182,9 +200,13 @@ impl Builder {
                      type_: None,
                  });
         body.extend(output_vars);
+        let ret_name = match trampoline.use_ret {
+            true => "retval",
+            false => "_",
+        };
```
Comment (EPashkin) [inline src/codegen/function_body_chunk.rs]: Please just "ret" and use if

**Comment by EPashkin** [comment]
Big thanks.
Looks good, I added small nits, will look thoroughly later.

File: src/analysis/bounds.rs (hunk 1 above)
Comment (MathieuDuponchelle) [reply src/analysis/bounds.rs]: done

File: src/codegen/function_body_chunk.rs (hunk 2 above)
Comment (MathieuDuponchelle) [reply src/codegen/function_body_chunk.rs]: done, though I'm not sure using if looks much better?

File: src/codegen/function_body_chunk.rs [hunk 3]
```diff
@@ -182,9 +200,15 @@ impl Builder {
                      type_: None,
                  });
         body.extend(output_vars);
+
+        let mut ret_name = "_";
+        if trampoline.use_ret {
+            ret_name = "ret";
+        }
```
Comment (EPashkin) [inline src/codegen/function_body_chunk.rs]: Why not `let ret_name = if trampoline.use_ret { "ret" } else { "_" };`

Comment (MathieuDuponchelle) [reply src/codegen/function_body_chunk.rs]: ah ! googled `let if else` but could only find `if let` documentation, tnx! (if you can't tell I'm not exactly fluent :)

File: src/codegen/function_body_chunk.rs [hunk 4]
```diff
@@ -159,6 +159,24 @@ impl Builder {
                     }
                 }
             }).collect();
+
+        if trampoline.use_ret {
+            if let Some(ref return_value) = trampoline.return_value {
+                let mem_mode = c_type_mem_mode_lib(env, return_value.typ, return_value.caller_allocates, return_value.transfer);
+                let value = Chunk::Name("retval".to_string());
```
Comment (EPashkin) [inline src/codegen/function_body_chunk.rs]: still `retval`

Comment (EPashkin) [inline src/codegen/function_body_chunk.rs]: Still "retval"

File: src/codegen/function_body_chunk.rs [hunk 5]
```diff
@@ -159,6 +159,24 @@ impl Builder {
                     }
                 }
             }).collect();
+
+        if trampoline.use_ret {
+            if let Some(ref return_value) = trampoline.return_value {
```
Comment (EPashkin) [inline src/codegen/function_body_chunk.rs]: Hm, there way to have use_ret=true and return_value = None ?
Maybe we can combine this fields to one return_value?

**Comment by EPashkin** [comment]
As I understand @sdroege already approved results.
@GuillaumeGomez You see any problem?

File: src/codegen/function_body_chunk.rs (hunk 5 above)
Comment (MathieuDuponchelle) [reply src/codegen/function_body_chunk.rs]: the opposite could be true yes (return_value that shouldn't be used), I could pass only the return_value when it needs to be used, but then that would make things a bit unintuitive.

**Comment by GuillaumeGomez** [comment]
I'm not much into async so apart from the code, I don't have much to say. So if it's good for you and @sdroege, it's good for me as well.

File: src/codegen/function_body_chunk.rs (hunk 5 above)
Comment (EPashkin) [reply src/codegen/function_body_chunk.rs]: As I know it normal in languages with optional types that if `return_value = Some(parameter)` then it used and `return_value = None` meant that nothing to return, so separate boolean field on the contrary makes it more confusing

**Comment by sdroege** [comment]
> As I understand @sdroege already approved results.

I only looked at the generated output, that seems great. And only shortly looked over this PR, nothing bad caught my eye. If the code generator changes look good to you, go for it :)

File: src/codegen/function_body_chunk.rs (hunk 5 above)
Comment (MathieuDuponchelle) [reply src/codegen/function_body_chunk.rs]: in that context, return_value is the return_value of the finish function, not something we have to "return", that's consistent with the other fields in the trampoline object (`output_params` in particular)

Comment (EPashkin) [reply src/codegen/function_body_chunk.rs]: Yes, from trampoline point of view it parameter for result for ffi-function (xxx_finish) and from other side it main type for good case. So IMHO we can use `ffi_ret` as field name, as I don't see normal way to separate parts of `Ok` content

Comment (MathieuDuponchelle) [reply src/codegen/function_body_chunk.rs]: I don't understand what you're saying here :)

Comment (EPashkin) [reply src/codegen/function_body_chunk.rs]: I says that this problem field in async trampoline (currently named `return_value`) is parameter that returns from FFI-function (xx_finish) in your case, so we can name it `ffi_ret` or `finish_ret` and use it in same way as now but without `use_ret` flag

Comment (EPashkin) [reply src/codegen/function_body_chunk.rs]: .. but set `ffi_ret` to `Some` only if `use_function_return_for_result` return true

Comment (MathieuDuponchelle) [reply src/codegen/function_body_chunk.rs]: Well if you insist on that I don't really mind, still don't like it though ;)

File: src/analysis/functions.rs [hunk 6]
```diff
@@ -192,9 +194,13 @@ fn analyze_function(
         if let Some((callback_type, bound_name)) = type_string {
             let func_name = func.c_identifier.as_ref().unwrap();
             let finish_func_name = replace_async_by_finish(func_name);
-            let rust_finish_func_name = replace_async_by_finish(&func.name);
             let mut output_params = vec![];
-            if let Some(function) = find_function(env, &rust_finish_func_name) {
+            let mut ffi_ret = None;
+            if let Some(function) = find_function(env, &finish_func_name) {
```
Comment (EPashkin) [inline src/analysis/functions.rs]: We still add trampoline when no finish function found, its normal?
If not then `output_params` and `ffi_ret` can be moved inside this if and created without `mut` modifier.

File: src/codegen/function_body_chunk.rs [hunk 7]
```diff
@@ -182,9 +198,12 @@ impl Builder {
                      type_: None,
                  });
         body.extend(output_vars);
+
+        let ret_name = if let Some(_) = trampoline.ffi_ret { "ret" } else { "_" };
```
Comment (EPashkin) [inline src/codegen/function_body_chunk.rs]: Better write `let ret_name = if trampoline.ffi_ret.is_some() { "ret" } else { "_" };`

File: src/analysis/functions.rs (hunk 6 above)
Comment (MathieuDuponchelle) [reply src/analysis/functions.rs]: Well I don't really know if that's normal, that's not really related to this PR though is it?

Comment (EPashkin) [reply src/analysis/functions.rs]: Yes

**Comment by EPashkin** [comment]
@MathieuDuponchelle Thanks, I merging this

**Comment by tmiasko** [comment]
Since this was merged I have panic when generating sys bindings: 
```
thread 'main' panicked at 'get rust type from return: Ignored("PixbufFormat")', libcore/result.rs:945:5
stack backtrace:
   0: std::sys::unix::backtrace::tracing::imp::unwind_backtrace
             at libstd/sys/unix/backtrace/tracing/gcc_s.rs:49
   1: std::sys_common::backtrace::_print
             at libstd/sys_common/backtrace.rs:71
   2: std::panicking::default_hook::{{closure}}
             at libstd/sys_common/backtrace.rs:59
             at libstd/panicking.rs:380
   3: std::panicking::default_hook
             at libstd/panicking.rs:396
   4: std::panicking::rust_panic_with_hook
             at libstd/panicking.rs:576
   5: std::panicking::begin_panic
             at libstd/panicking.rs:537
   6: std::panicking::begin_panic_fmt
             at libstd/panicking.rs:521
   7: rust_begin_unwind
             at libstd/panicking.rs:497
   8: core::panicking::panic_fmt
             at libcore/panicking.rs:71
   9: core::result::unwrap_failed
  10: gir::analysis::bounds::Bounds::add_for_parameter
  11: gir::analysis::functions::analyze_function
  12: gir::analysis::functions::analyze
  13: gir::analysis::object::class
  14: gir::analysis::run
  15: gir::do_main
  16: gir::main
  17: std::rt::lang_start::{{closure}}
  18: std::panicking::try::do_call
             at libstd/rt.rs:59
             at libstd/panicking.rs:479
  19: __rust_maybe_catch_panic
             at libpanic_unwind/lib.rs:102
  20: std::rt::lang_start_internal
             at libstd/panicking.rs:458
             at libstd/panic.rs:358
             at libstd/rt.rs:58
  21: main
  22: __libc_start_main
  23: _start
```

**Pull request merged**